src/geocel/rasterize/Raytracer.hh

```diff
@@ -0,0 +1,182 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file geocel/rasterize/Raytracer.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/math/Algorithms.hh"
+
+#include "ImageLineView.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Trace each pixel along a line.
+ *
+ * \tparam GTV Geometry Track View
+ * \tparam F   Calculate the pixel value (result_type) given GTV
+ */
+template<class GTV, class F>
+class Raytracer
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using result_type = int;
+    //!@}
+
+  public:
+    // Construct from geo track view, function to calculate ID, and image
+    inline CELER_FUNCTION
+    Raytracer(GTV&& geo, F&& calc_id, ImageLineView const& image);
+
+    // Calculate the value for the next pixel
+    inline CELER_FUNCTION result_type operator()(size_type pix);
+
+  private:
+    //// DATA ////
+
+    GTV geo_;
+    F calc_id_;
+    ImageLineView const& image_;
+
+    size_type pixel_;
+
+    //// HELPER FUNCTIONS ////
+
+    // Initialize the geometry at this pixel index
+    inline CELER_FUNCTION void initialize_at_pixel(size_type pix);
+
+    //! Sentinel value for needing the pixel to be reset
+    static CELER_CONSTEXPR_FUNCTION size_type invalid_pixel()
+    {
+        return static_cast<size_type>(-1);
+    }
+
+    //! Stop tracking the pixel if this many crossings occur
+    static CELER_CONSTEXPR_FUNCTION short int max_crossings_per_pixel()
+    {
+        return 32;
+    }
+};
+
+//---------------------------------------------------------------------------//
+// DEDUCTION GUIDES
+//---------------------------------------------------------------------------//
+
+template<class GTV, class F>
+CELER_FUNCTION Raytracer(GTV&&, F&&, ImageLineView const&)
+    -> Raytracer<GTV, F>;
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from geo track view, function to calculate ID, and image.
+ */
+template<class GTV, class F>
+CELER_FUNCTION
+Raytracer<GTV, F>::Raytracer(GTV&& geo, F&& calc_id, ImageLineView const& image)
+    : geo_{celeritas::forward<GTV>(geo)}
+    , calc_id_{celeritas::forward<F>(calc_id)}
```

Comment:
For a real GTV (e.g. in the case of VecGeom or ORANGE), is this `calc_id` supposed to be the geometry engine's locate function, like VecGeom's `Locate()`?  Where is the similar ORANGE function?

Reply:
So this flexibility is helpful to define colors in different ways to paint components in rasterization pictures?
Nice!